product/scanning.py

```diff
@@ -118,22 +124,41 @@ def upload_image(
 
         if scan_result:
             # check if image version changed
-
             metadata = scan_result.custom_data()
             image_reference = metadata.get('image-reference')
-            image_changed = image_reference == container_image.image_reference()
+            image_changed = image_reference != container_image.image_reference()
 
-            if not image_changed:
-                # image reference did not change - early exit
+            if not image_changed and upload_mode is UploadMode.UPDATE:
+                # image reference did not change and we need not rescan - early exit
                 return upload_result(
                     status=UploadStatus.SKIPPED_ALREADY_EXISTED,
                     result=scan_result,
                 )
-            else:
-                pass # continue with regular image upload; overwrite should take place
-                # due to identical name (hopefully)
 
-        # image was not yet uploaded - do this now
+            if upload_mode is upload_mode.RESCAN:
+                # check if rescan can be triggered and whether it makes sense
+                # to do so (according to protecode-response)
+                scan_result_short = self._api.scan_result_short(scan_result.product_id())
+                if scan_result_short.is_stale():
+                    if scan_result_short.has_binary():
+                        # The binary is still available remote, just trigger a rescan
+                        self._api.rescan(scan_result.product_id())
+                        status, result = self.wait_and_retrieve_upload_result(
+                            wait_for_result,
+                            scan_result.product_id(),
+                            enum_type=ScanStatus,
+                        )
+                        return upload_result(
+                            status = status,
+                            result = result,
+                        )
+                else:
```

Comment:
explicitly check for the other upload modes; add another else that will raise NotImplemented to detect missing case handling